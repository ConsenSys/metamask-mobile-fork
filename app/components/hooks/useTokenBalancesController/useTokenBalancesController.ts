import { useSelector } from 'react-redux';
import { EngineState } from '../../../selectors/types';
import { isEqual } from 'lodash';
import { ControllerHookType } from '../controllerHook.types';

const useTokenBalancesController = (): ControllerHookType => {
  const tokenBalances = useSelector(
    (state: EngineState) =>
      state.engine.backgroundState?.TokenBalancesController?.contractBalances,
    isEqual,
  );

  return { data: tokenBalances };
};

export default useTokenBalancesController;
