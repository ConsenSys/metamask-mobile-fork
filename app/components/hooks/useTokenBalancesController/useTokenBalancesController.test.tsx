import React, { useEffect } from 'react';
import { Text } from 'react-native';
import { Provider } from 'react-redux';
import { createStore } from 'redux';
import { act, render, waitFor } from '@testing-library/react-native';
import useTokenBalancesController from './useTokenBalancesController';
import { BN } from 'ethereumjs-util';

// initial state for the test store
const initialState = {
  engine: {
    backgroundState: {
      TokenBalancesController: {
        contractBalances: {
          '0x326836cc6cd09B5aa59B81A7F72F25FcC0136b95': new BN(0x2a),
        },
      },
    },
  },
};

// test reducer for the test store
const testBalancesReducer = (state: any, action: any) => {
  if (action.type === 'add-balances') {
    const newBalances = {
      ...state.engine.backgroundState.TokenBalancesController.contractBalances,
      ...action.value,
    };
    state.engine.backgroundState.TokenBalancesController.contractBalances =
      newBalances;
    return state;
  }
  return initialState;
};

// create a test store, not a mock, as we need to test for content changes
const testStore = createStore(testBalancesReducer);

const Wrapper = ({ children }: any) => (
  <Provider store={testStore}>{children}</Provider>
);

const spyOnDummyUseEffect = jest.fn();
// Dummy component to test the useTokenBalancesController hook
// and be able to detect the rerenders using the useEffect spy
const DummyTestComponent = () => {
  const balances = useTokenBalancesController();

  useEffect(() => {
    spyOnDummyUseEffect();
  }, [balances]);

  // render the balances as a string to be able to compare the snapshots
  return <Text>{JSON.stringify(balances)}</Text>;
};

describe('useTokenBalancesController()', () => {
  it('should only rerender when state changed', async () => {
    // render the component with the test store
    const { toJSON } = render(<DummyTestComponent />, { wrapper: Wrapper });

    // check the useEffect is called once and initial balances are returned
    await waitFor(async () => {
      // check that the initial balances are returned
      expect(spyOnDummyUseEffect).toHaveBeenCalledTimes(1);
      expect(toJSON()).toMatchSnapshot();
    });

    // add a new balance to the store
    await act(async () => {
      testStore.dispatch({
        type: 'add-balances',
        value: {
          '0x326836cc6cd09B5aa59B81A7F72F25FcC0136b96': new BN(0x539),
        },
      });
    });

    // check that the useEffect is called again and the new updated balances are returned
    await waitFor(() => {
      expect(spyOnDummyUseEffect).toHaveBeenCalledTimes(2);
      expect(toJSON()).toMatchSnapshot();
    });

    // add the same balance to the store again, should override the previous one
    await act(async () => {
      testStore.dispatch({
        type: 'add-balances',
        value: {
          '0x326836cc6cd09B5aa59B81A7F72F25FcC0136b96': new BN(0x539),
        },
      });
    });

    // check that the useEffect is not called again and the returned balances are unchanged
    await waitFor(async () => {
      expect(spyOnDummyUseEffect).toHaveBeenCalledTimes(2);
      expect(toJSON()).toMatchSnapshot();
    });
  });
});
