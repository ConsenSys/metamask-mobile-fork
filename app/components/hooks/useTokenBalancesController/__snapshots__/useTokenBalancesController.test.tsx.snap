// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useTokenBalancesController() should only rerender when state changed 1`] = `
<Text>
  {"data":{"0x326836cc6cd09B5aa59B81A7F72F25FcC0136b95":"2a"}}
</Text>
`;

exports[`useTokenBalancesController() should only rerender when state changed 2`] = `
<Text>
  {"data":{"0x326836cc6cd09B5aa59B81A7F72F25FcC0136b95":"2a","0x326836cc6cd09B5aa59B81A7F72F25FcC0136b96":"539"}}
</Text>
`;

exports[`useTokenBalancesController() should only rerender when state changed 3`] = `
<Text>
  {"data":{"0x326836cc6cd09B5aa59B81A7F72F25FcC0136b95":"2a","0x326836cc6cd09B5aa59B81A7F72F25FcC0136b96":"539"}}
</Text>
`;
